src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs

```diff
@@ -0,0 +1,196 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Helper type for building dynamically-sized arrays while minimizing allocations and copying.
+    /// </summary>
+    internal struct LargeArrayBuilder<T>
+    {
+        private const int StartingCapacity = 4;
+        private const int ResizeLimit = 32;
+
+        private T[] _first;                 // The first buffer we store items in. Resized until ResizeLimit.
+        private ArrayBuilder<T[]> _buffers; // After ResizeLimit * 2, we store previous buffers we've filled out here.
+        private T[] _current;               // Current buffer we're reading into. Is _first <= ResizeLimit.
+        private int _index;                 // Index into the current buffer.
+        private int _count;                 // Count of all of the items in this builder.
+
+        /// <summary>
+        /// Constructs a new builder.
+        /// </summary>
+        /// <param name="initialize">Pass <c>true</c>.</param>
+        public LargeArrayBuilder(bool initialize) : this()
```

Comment:
@stephentoub 

> it looks to me the same as having a bunch of fields in a class that are explicitly initialized to default(T) at the field declaration site and then subsequently overwritten in an explicit ctor

If you think about it, that can be faster since the zeroing can be done in bulk vs. initializing fields 1 at a time. For example, on my machine the jit generates code to use `ymm` registers:

```asm
       C4E17957C0           vxorpd   ymm0, ymm0
       C4E17A7F02           vmovdqu  qword ptr [rdx], ymm0
       C4E17A7F4210         vmovdqu  qword ptr [rdx+16], ymm0
```

> If there's actually an inlining difference and it matters, that's something that should be addressed in the JIT.

Considering things from the viewpoint of the JIT, isn't the constructor just a bunch of arbitrary code? So then when it looks at the constructor and sees that it has a huge body, it might not decide to inline even though all of the code is just default-initializing the struct. It would seem that if the JIT were to try to inline this, it would waste time for other cases where that arbitrary code isn't all just default-initialization.

I'm not really familiar with how the JIT works, but just a guess... cc @mikedn, @benaadams